src/policy/ancestor_packages.h

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) 2021-2022 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_POLICY_ANCESTOR_PACKAGES_H
+#define BITCOIN_POLICY_ANCESTOR_PACKAGES_H
+
+#include <policy/feerate.h>
+#include <policy/packages.h>
+
+#include <vector>
+
+/** A potential BIP331 Ancestor Package, i.e. one transaction with its set of ancestors.
+ * This class does not have any knowledge of chainstate, so it cannot determine whether all
+ * unconfirmed ancestors are present. Its constructor accepts any list of transactions that
```

Comment:
So if my understanding of the API is correct, you receive a package from `AcceptPackage()`, and there is a call to the constructor `AncestorPackage()`, then we verify each transaction being component of the package is in the mempool by wtxid or txid or `PreChecks()` them.

If the transaction is `PreChecks()` valid, then we call `AddFeeAndVsize()`. If we have a `TX_SINGLE_FAILURE`, we add `AddFeeAndVsize()` or if we have a `TX_MISSING_INPUTS`, we `SkipWithDescendants()` the transaction. All other cases are treaded as impossible or failure.

So when we call `LinearizeWithFees()` in `AcceptPackage` and therefore start to rely on `MiniMiner` calculations. we have asserted than all packages component linearize are either spending a UTXO in the chainstate set, or spending an in-package component, so the worst-case scenario from a DoS viewpoint will be the limits as set in BIP331 / nversion=3 ?

Reply:
> So if my understanding of the API is correct, you receive a package from AcceptPackage() ...

Correct

> we have asserted than all packages component linearize are either spending a UTXO in the chainstate set, or spending an in-package component, so the worst-case scenario from a DoS viewpoint will be the limits as set in BIP331 / nversion=3 ?

The first check in `AcceptPackage` is `IsPackageWellFormed` which requires the package to be 25 transactions at most. The linearization with `MiniMiner` also does not fetch in-mempool transactions. So you can assume we never construct the miniminer with more than 25 transactions.

This is independent of v3 so those limits aren't enforced.

BIP331 doesn't impose a maximum size other than 100 for pkgtxns, but I expect the p2p logic to observe that an `ancpkginfo` with more than 25 transactions wouldn't be submitted and just exit there without downloading the transactions. So we wouldn't send a `getpkgtxns` request for more than 25, and thus would immediately reject a (unsolicited) `pkgtxns` with more than 25. See logic for handling `ancpkginfo`, we will give up on downloading an ancestor package from a peer if it has too many transactions: https://github.com/bitcoin/bitcoin/pull/27742/files#diff-6875de769e90cec84d2e8a9c1b962cdbcda44d870d42e4215827e599e11e90e3R4398.

Reply:
> So you can assume we never construct the miniminer with more than 25 transactions.

Thanks, yes I think `MAX_PACKAGE_COUNT` as enforced by `IsPackageWellFormed` is the upper bound that one can assume in the calculation of `MiniMiner`. This can ease the DoS reasoning to link the max `ancpkginfo` issued at the p2p level with `MAX_PACKAGE_COUNT` as requested by the mempool logic to save on bandwidth both by the local node (the list of wtxids announced) and its peer (`pkgtnxs`’s txn), I think.